lightning-invoice/src/utils.rs

```diff
@@ -510,6 +540,60 @@ where
 	}
 }
 
+
+/// Used to store information about all the HTLCs that are inflight across all payment attempts
+pub(crate) struct AccountForInFlightHtlcs<'a, S: Score> {
+	scorer: &'a mut S,
+	/// Maps a channel's short channel id and its direction to the liquidity used up.
+	inflight_htlcs: HashMap<(u64, bool), u64>,
+}
+
+impl<'a, S: Score> AccountForInFlightHtlcs<'a, S> {
+	pub(crate) fn new(scorer: &'a mut S, inflight_htlcs: HashMap<(u64, bool), u64>) -> Self {
+		AccountForInFlightHtlcs {
+			scorer,
+			inflight_htlcs
+		}
+	}
+}
+
+#[cfg(c_bindings)]
+impl<'a, S:Score> lightning::util::ser::Writeable for AccountForInFlightHtlcs<'a, S> {
+	fn write<W: lightning::util::ser::Writer>(&self, writer: &mut W) -> Result<(), std::io::Error> { self.scorer.write(writer) }
+}
+
+impl<'a, S: Score> Score for AccountForInFlightHtlcs<'a, S> {
+	fn channel_penalty_msat(&self, short_channel_id: u64, source: &NodeId, target: &NodeId, usage: ChannelUsage) -> u64 {
+		if let Some(used_liqudity) = self.inflight_htlcs.get(&(short_channel_id, source < target)) {
+			let usage = ChannelUsage {
+				inflight_htlc_msat: usage.inflight_htlc_msat + used_liqudity,
+				..usage
+			};
+
+			self.scorer.channel_penalty_msat(short_channel_id, source, target, usage)
+		} else {
+			self.scorer.channel_penalty_msat(short_channel_id, source, target, usage)
+		}
+	}
+
+	fn payment_path_failed(&mut self, path: &[&RouteHop], short_channel_id: u64) {
+		self.scorer.payment_path_failed(path, short_channel_id)
+	}
+
+	fn payment_path_successful(&mut self, path: &[&RouteHop]) {
+		self.scorer.payment_path_successful(path)
+	}
+
+	fn probe_failed(&mut self, path: &[&RouteHop], short_channel_id: u64) {
+		self.scorer.probe_failed(path, short_channel_id)
+	}
+
+	fn probe_successful(&mut self, path: &[&RouteHop]) {
+		self.scorer.probe_successful(path)
+	}
```

Comment:
These are never actually called since `AccountForInFlightHtlcs` is only used for `channel_penalty_msat`, so they can be implemented as `unreachable!()`. I wonder if we should split `Score` into two different traits or if we are fine with duplicating part of the interface in `Router`. I guess either way, any `Score` implementation used by `DefaultRouter` would also need to implement this half of the trait. Not sure what would be prefered. @tnull @TheBlueMatt Any thoughts?

Reply:
Addressed in 7d95196a